src/modules/mavlink/mavlink_messages.cpp

```diff
@@ -3955,186 +3967,599 @@ class MavlinkStreamHighLatency : public MavlinkStream
+			    ((status.onboard_control_sensors_enabled & MAV_SYS_STATUS_SENSOR_3D_ACCEL2) &&
+			     !(status.onboard_control_sensors_health & MAV_SYS_STATUS_SENSOR_3D_ACCEL2))) {
+				msg->failure_flags |= HL_FAILURE_FLAG_3D_ACCEL;
+			}
+
+			if (((status.onboard_control_sensors_enabled & MAV_SYS_STATUS_SENSOR_3D_GYRO)
+			     && !(status.onboard_control_sensors_health & MAV_SYS_STATUS_SENSOR_3D_GYRO)) ||
+			    ((status.onboard_control_sensors_enabled & MAV_SYS_STATUS_SENSOR_3D_GYRO2) &&
+			     !(status.onboard_control_sensors_health & MAV_SYS_STATUS_SENSOR_3D_GYRO2))) {
+				msg->failure_flags |= HL_FAILURE_FLAG_3D_GYRO;
+			}
 
-			mavlink_msg_high_latency_send_struct(_mavlink->get_channel(), &msg);
+			if (((status.onboard_control_sensors_enabled & MAV_SYS_STATUS_SENSOR_3D_MAG)
+			     && !(status.onboard_control_sensors_health & MAV_SYS_STATUS_SENSOR_3D_MAG)) ||
+			    ((status.onboard_control_sensors_enabled & MAV_SYS_STATUS_SENSOR_3D_MAG2) &&
+			     !(status.onboard_control_sensors_health & MAV_SYS_STATUS_SENSOR_3D_MAG2))) {
+				msg->failure_flags |= HL_FAILURE_FLAG_3D_MAG;
+			}
+
+			if ((status.onboard_control_sensors_enabled & MAV_SYS_STATUS_TERRAIN)
+			    && !(status.onboard_control_sensors_health & MAV_SYS_STATUS_TERRAIN)) {
+				msg->failure_flags |= HL_FAILURE_FLAG_TERRAIN;
+			}
+
+			if (status.rc_signal_lost) {
+				msg->failure_flags |= HL_FAILURE_FLAG_RC_RECEIVER;
+			}
+
+			if (status.engine_failure) {
+				msg->failure_flags |= HL_FAILURE_FLAG_ENGINE;
+			}
+
+			if (status.mission_failure) {
+				msg->failure_flags |= HL_FAILURE_FLAG_MISSION;
+			}
+
+			// flight mode
+			union px4_custom_mode custom_mode;
+			uint8_t mavlink_base_mode;
+			get_mavlink_custom_mode(&status, &mavlink_base_mode, &custom_mode);
+			msg->custom_mode = custom_mode.custom_mode_hl;
+		}
+
+		return updated;
+	}
+
+	bool write_vehicle_status_flags(mavlink_high_latency2_t *msg)
+	{
+		struct vehicle_status_flags_s status_flags = {};
+
+		const bool updated = _status_flags_sub->update(&_status_flags_time, &status_flags);
+
+		if (_status_flags_time > 0) {
+			if (!status_flags.condition_global_position_valid) { //TODO check if there is a better way to get only GPS failure
+				msg->failure_flags |= HL_FAILURE_FLAG_GPS;
+			}
+
+			if (status_flags.offboard_control_signal_lost && status_flags.offboard_control_signal_found_once) {
+				msg->failure_flags |= HL_FAILURE_FLAG_OFFBOARD_LINK;
+			}
+		}
+
+		return updated;
+	}
+
+	bool write_wind_estimate(mavlink_high_latency2_t *msg)
+	{
+		struct wind_estimate_s wind = {};
+
+		const bool updated = _wind_sub->update(&_wind_time, &wind);
+
+		if (_wind_time > 0) {
+			msg->wind_heading = static_cast<uint8_t>(
+						    math::degrees(_wrap_2pi(atan2f(wind.windspeed_east, wind.windspeed_north))) * 0.5f);
 		}
 
 		return updated;
 	}
+
+	void update_data()
+	{
+		const float update_rate = 1.0f / (_mavlink->get_main_loop_delay() * 1e-6f);
```

Comment:
I'm not sure if you can rely on this being the actual execution rate.

Reply:
True, I checked it and the actual update rate is much lower. I changed it now to using the filtered update rate computed from the time elapsed. I am using the filtered rate instead of the actual one as the rate is only used to determine when switching from averaging to the moving average filter and therefore using the actual rate would case switching between the methods too much.